File: common/src/main/java/cloudcmd/common/FileMetaData.java
package cloudcmd.common;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Set;

public class FileMetaData
{
  private JSONObject _data;
  private String _hash;

  public JSONArray getBlockHashes()
      throws JSONException
  {
    return _data.getJSONArray("blocks");
  }

  public Set<String> getTags()
      throws JSONException
  {
    return JsonUtil.createSet(_data.getJSONArray("tags"));
  }

  public JSONObject toJson()
      throws JSONException
  {
    JSONObject obj = new JSONObject();
    obj.put("hash", _hash);
    obj.put("data", _data);
    return obj;
  }

  public static FileMetaData create(JSONObject jsonObject)
      throws IOException
  {
    return
      create(
        CryptoUtil.computeHashAsString(new ByteArrayInputStream(jsonObject.toString().getBytes("UTF-8"))) + ".meta",
        jsonObject);
  }

  public static FileMetaData create(String hash, JSONObject data)
  {
    FileMetaData meta = new FileMetaData();
    meta._data = data;
    meta._hash = hash;
    return meta;
  }

  public String getHash()
  {
    return _hash;
  }

  public String getParent()
      throws JSONException
  {
    return _data.has("parent") ? _data.getString("parent") : null;
  }

  public String getPath()
      throws JSONException
  {
    return _data.getString("path");
  }

  public String getFilename()
      throws JSONException
  {
    return _data.getString("filename");
  }

  public String getFileExt()
      throws JSONException
  {
    return _data.has("fileext") ? _data.getString("fileext") : null;
  }

  public Long getFileDate()
      throws JSONException
  {
    return _data.getLong("filedate");
  }

  public Long getFileSize()
      throws JSONException
  {
    return _data.getLong("filesize");
  }

  public String getDataAsString()
  {
    return _data.toString();
  }
}

File: common/src/main/scala/cloudcmd/common/util/FileMetaData.scala
package cloudcmd.common.util

import java.io.ByteArrayInputStream
import org.json.JSONArray
import org.json.JSONObject

object FileMetaData {
  def create(jsonObject: JSONObject): FileMetaData = {
    create(CryptoUtil.computeHashAsString(new ByteArrayInputStream(jsonObject.toString.getBytes("UTF-8"))) + ".meta", jsonObject)
  }

  def create(hash: String, data: JSONObject): FileMetaData = {
    val meta: FileMetaData = new FileMetaData
    meta._data = data
    meta._hash = hash
    meta
  }
}

class FileMetaData {
  override def hashCode: Int = {
    _hash.hashCode
  }

  override def equals(other: Any): Boolean = {
    _hash.equals(other)
  }

  def getBlockHashes: JSONArray = {
    _data.getJSONArray("blocks")
  }

  def getTags: Set[String] = {
    JsonUtil.createSet(_data.getJSONArray("tags"))
  }

  def toJson: JSONObject = {
    val obj: JSONObject = new JSONObject
    obj.put("hash", _hash)
    obj.put("data", _data)
    obj
  }

  def getHash: String = {
    _hash
  }

  def getParent: String = {
    if (_data.has("parent")) _data.getString("parent") else null
  }

  def getPath: String = {
    _data.getString("path")
  }

  def getFilename: String = {
    _data.getString("filename")
  }

  def getFileExt: String = {
    if (_data.has("fileext")) _data.getString("fileext") else null
  }

  def getFileDate: Long = {
    _data.getLong("filedate")
  }

  def getFileSize: Long = {
    _data.getLong("filesize")
  }

  def getRawData: JSONObject = {
    _data
  }

  def getDataAsString: String = {
    _data.toString
  }

  private var _data: JSONObject = null
  private var _hash: String = null
}